source/openqasm/openqasm3/ast.py

```diff
@@ -1004,6 +1005,26 @@ class ForInLoop(Statement):
     block: List[Statement]
 
 
+@dataclass
+class SwitchStatement(Statement):
+    """A switch-case statement.
+
+    The literal cases are stored in the `cases` dictionary, in declaration
+    order.  Python's `dict` guarantees (for all supported versions of Python)
+    that the iteration order will be insertion order, so this field *is*
+    ordered.
+
+    The default case, if any, is in the `default` attribute.
+    """
+
+    target: Expression
+    cases: Dict[Tuple[int, ...], CompoundStatement]
```

Comment:
I'd read "integer constant" to mean specifically "integer literal", mostly because there weren't any examples in the text of using constant identifiers, but I think you're right - "constant expression" is the same terminology the C99 spec uses, and `1 + 2` is a valid label in C.  It might be worth adding some examples of constant expressions to the spec text to make it clear as well.

A list of pairs is completely fine for the AST, imo.  I made it a `dict` originally only because it was a minor convenience without sacrifice in Python 3.7+, since `dict` retains insertion order, so it's a proxy for a list.  It's no trouble to make it only a list, and I don't think that anything operating on the AST will really care.

Modified in ceac60f.  As a side effect, we largely have to lose the duplicate-case detection at this point in AST generation, since we can only reliably do that after constant-propagation, which this generator doesn't do.